requests-for-comments/2022-11-02-data-exploration.md

```diff
@@ -1,34 +1,118 @@
+# Request for comments: Data exploration: Nail Data Tables + V2 of the Insights API - Marius Andra
 
-This is a proposal to change the way we query for data in PostHog, and how we display the results.
+## Summary
 
-## Why?
+Problem 1 (The user's problem): PostHog is good for the creator (you know what you what insight you want to create) and consumer (viewing already made insights). But less so for the explorer (where the questions are higher level and more uncertain). For these questions you need to be able to "drill down" into the specific data and roll-up into more aggregated views.
 
-With PostHog, users own their data. Yet when exploring it, they're limited to the tools we've built. Sometimes they want more:
+Problem 2 (The engineering problem): Our API for querying insights has became very complex. It's slowing down our ability to iterate. Instead of extending our existing API we should be a more powerful V2 Insights API.
+    - Why extend our API now rather than refactor later on? Many of the important feature requests will make the API even more complex. We should therefore refactor the API now before implementing the requested feature requests.
 
-- Insight person modals limit you to looking at a list of people. You can't drill them further (e.g. add a filter, break down by cohorts, etc).
-- The live events table supports event properties as columns, but that's it. What about person properties? What about aggregations? Sorting is not supported.
-- You can't have insights with non-event data sources (_persons_, event names?), even though some chart types could support them (e.g. pie chart of all persons in the system broken down by country)
-- Persons tables are very basic. No column customisation. No math or formulas.
-- There are charts in insights that can't be added to dashboards (e.g. retention line chart, funnel correlation results).
-- The listing pages (events, persons, etc) support filters, but those can't be saved.
-- There are parameters you can't directly filter for or list (such as `distinct_id`, `timestamp`).
-- Our `FilterType` has been append-only for 2+ years and is long overdue for a serious cleanup.
-- Multiple endpoints accept various types of filters with various capabilities (`/insights/funnel/correlations` vs `/insights/trend/` vs `/events` vs `/persons/trends`). They're all typed by this same object. It's impossible to type the outputs without blindly guessing.
+Proposed solution: **Nail Data Tables for exploration**.
 
-We have an opportunity to build a unified "data exploration" view, which would solve all or most of the above, and open the floodgates to allowing any type of analysis.
+This has two parts:
 
-The proposed changes below will and also unblock:
+1. A single table which lets you flip between persons, events and recordings (need a better name, currently called Drill Down view OR Universal Search)
+2. Pivot tables for viewing aggregate level statistics.
```

Comment:
Note: I think adding "pivoting" e.g. grouping/aggregating to the project instead of just filtering expands the scope roughly 3x - we do end up basically needing to rewrite $ALL_OF_POSTHOG_QUERIES and I can see this taking months up to a year depending _while slowing down all other developments to a crawl_.

Scary!

Reply:
Discussed this in a bit more depth at a standup. The solution is cutting scope a bit in terms of building things: I think it might be sensible to:

1. First build pivot table for only a single view
2. Delay building "advanced filtering" (e.g. allowing filtering to show users who dropped off at this funnel step 3)